File: core/protocol/src/commonMain/kotlin/dev/audiocompanion/protocol/Wire.kt
package dev.audiocompanion.protocol

/**
 * Little-endian packed-struct primitives (spec Section 2). Shared by the message
 * codecs here and by :core:storage, whose frame-log records deliberately use the
 * same record shape as STREAM_DATA frame entries.
 */
class WireWriter(initialCapacity: Int = 32) {
    private var buf = ByteArray(initialCapacity)
    private var size = 0

    private fun ensure(extra: Int) {
        if (size + extra > buf.size) {
            buf = buf.copyOf(maxOf(buf.size * 2, size + extra))
        }
    }

    fun u8(value: Int): WireWriter {
        ensure(1)
        buf[size++] = (value and 0xFF).toByte()
        return this
    }

    fun u16(value: Int): WireWriter {
        ensure(2)
        buf[size++] = (value and 0xFF).toByte()
        buf[size++] = ((value ushr 8) and 0xFF).toByte()
        return this
    }

    fun u32(value: UInt): WireWriter {
        ensure(4)
        buf[size++] = (value and 0xFFu).toByte()
        buf[size++] = ((value shr 8) and 0xFFu).toByte()
        buf[size++] = ((value shr 16) and 0xFFu).toByte()
        buf[size++] = ((value shr 24) and 0xFFu).toByte()
        return this
    }

    fun u64(value: ULong): WireWriter {
        ensure(8)
        var v = value
        repeat(8) {
            buf[size++] = (v and 0xFFu).toByte()
            v = v shr 8
        }
        return this
    }

    fun bytes(value: ByteArray): WireWriter {
        ensure(value.size)
        value.copyInto(buf, size)
        size += value.size
        return this
    }

    fun toByteArray(): ByteArray = buf.copyOf(size)
}

/** Reader over a ByteArray. Callers must check [remaining] before reading. */
class WireReader(private val bytes: ByteArray, private var offset: Int = 0) {
    val remaining: Int get() = bytes.size - offset

    fun u8(): Int = bytes[offset++].toInt() and 0xFF

    fun u16(): Int {
        val v = (bytes[offset].toInt() and 0xFF) or
            ((bytes[offset + 1].toInt() and 0xFF) shl 8)
        offset += 2
        return v
    }

    fun u32(): UInt {
        var v = 0u
        for (i in 3 downTo 0) {
            v = (v shl 8) or (bytes[offset + i].toUInt() and 0xFFu)
        }
        offset += 4
        return v
    }

    fun u64(): ULong {
        var v = 0uL
        for (i in 7 downTo 0) {
            v = (v shl 8) or (bytes[offset + i].toULong() and 0xFFu)
        }
        offset += 8
        return v
    }

    fun bytes(count: Int): ByteArray {
        val out = bytes.copyOfRange(offset, offset + count)
        offset += count
        return out
    }
}

File: ios/PebbleAudioKit/Sources/WireProtocol/Wire.swift
import Foundation

/// Little-endian packed-struct primitives (spec Section 2). Shared by the message codecs
/// and by SegmentStore, whose frame-log records deliberately use the same record shape as
/// STREAM_DATA frame entries. Port of `core/protocol/.../Wire.kt`.
public struct WireWriter {
    private var buf: [UInt8]

    public init(initialCapacity: Int = 32) {
        buf = []
        buf.reserveCapacity(initialCapacity)
    }

    @discardableResult
    public mutating func u8(_ value: Int) -> WireWriter {
        buf.append(UInt8(truncatingIfNeeded: value))
        return self
    }

    @discardableResult
    public mutating func u16(_ value: Int) -> WireWriter {
        buf.append(UInt8(truncatingIfNeeded: value))
        buf.append(UInt8(truncatingIfNeeded: value >> 8))
        return self
    }

    @discardableResult
    public mutating func u32(_ value: UInt32) -> WireWriter {
        buf.append(UInt8(truncatingIfNeeded: value))
        buf.append(UInt8(truncatingIfNeeded: value >> 8))
        buf.append(UInt8(truncatingIfNeeded: value >> 16))
        buf.append(UInt8(truncatingIfNeeded: value >> 24))
        return self
    }

    @discardableResult
    public mutating func u64(_ value: UInt64) -> WireWriter {
        var v = value
        for _ in 0..<8 {
            buf.append(UInt8(truncatingIfNeeded: v))
            v >>= 8
        }
        return self
    }

    @discardableResult
    public mutating func bytes(_ value: [UInt8]) -> WireWriter {
        buf.append(contentsOf: value)
        return self
    }

    @discardableResult
    public mutating func bytes(_ value: Data) -> WireWriter {
        buf.append(contentsOf: value)
        return self
    }

    public func toBytes() -> [UInt8] { buf }
    public func toData() -> Data { Data(buf) }
}

/// Reader over a byte buffer. Callers must check `remaining` before reading.
public struct WireReader {
    private let bytes: [UInt8]
    private var offset: Int

    public init(_ bytes: [UInt8], offset: Int = 0) {
        self.bytes = bytes
        self.offset = offset
    }

    public init(_ data: Data, offset: Int = 0) {
        self.bytes = [UInt8](data)
        self.offset = offset
    }

    public var remaining: Int { bytes.count - offset }

    public mutating func u8() -> Int {
        defer { offset += 1 }
        return Int(bytes[offset])
    }

    public mutating func u16() -> Int {
        defer { offset += 2 }
        return Int(bytes[offset]) | (Int(bytes[offset + 1]) << 8)
    }

    public mutating func u32() -> UInt32 {
        var v: UInt32 = 0
        for i in stride(from: 3, through: 0, by: -1) {
            v = (v << 8) | UInt32(bytes[offset + i])
        }
        offset += 4
        return v
    }

    public mutating func u64() -> UInt64 {
        var v: UInt64 = 0
        for i in stride(from: 7, through: 0, by: -1) {
            v = (v << 8) | UInt64(bytes[offset + i])
        }
        offset += 8
        return v
    }

    public mutating func readBytes(_ count: Int) -> [UInt8] {
        defer { offset += count }
        return Array(bytes[offset..<(offset + count)])
    }
}
